unboundid-provider/src/main/java/org/ldaptive/provider/unboundid/UnboundIDProvider.java

```diff
@@ -83,6 +110,43 @@ protected LDAPConnectionOptions getDefaultLDAPConnectionOptions(final Connection
     if (cc.getResponseTimeout() != null) {
       options.setResponseTimeoutMillis(cc.getResponseTimeout().toMillis());
     }
+    // UnboundID does not do hostname verification by default
+    // set a default hostname verifier if no SSL socket verifier configured
+    final HostnameVerifier verifier;
+    if (cc.getSslConfig() != null && !cc.getSslConfig().isEmpty()) {
+      if (cc.getSslConfig().getHostnameVerifier() != null) {
+        verifier = new HostnameVerifierAdapter(cc.getSslConfig().getHostnameVerifier());
+      } else {
+        verifier = new DefaultHostnameVerifier();
+      }
+    } else {
+      verifier = new DefaultHostnameVerifier();
+    }
+    options.setSSLSocketVerifier(new SSLSocketVerifier()
+    {
+      /** Logger for this class. */
+      private  final Logger logger = LoggerFactory.getLogger(getClass());
+
+
+      @Override
+      public void verifySSLSocket(
+        final String host,
+        final int port,
+        final SSLSocket sslSocket)
+        throws LDAPException
+      {
+        logger.trace("Verifying SSLSocket {} for host {} with verifier {}", sslSocket, host, verifier);
+        try {
+          sslSocket.getSession().getPeerCertificates();
```

Comment:
This raised an eyebrow. It's clear you need an `SSLSession` in order to perform verification, but I would expect you to have a local variable that stores the reference you get from this operation. Then it would be perfectly clear why this operation precedes the subsequent hostname verification.

Reply:
It's the call to getPeerCertificates that throws, not getSession.
I'm not sure if a local variable makes that any clearer.